### ISSUE TITLE
Allow dependencies that only apply to the bin target

### ISSUE THREAD

**Issue by tbu-** [rust-lang/cargo#1982]
Use case: I have a library that also produces an executable. The executable needs `env_logger` in order to be able to log, the library shouldn't carry that dependency though, as it pulls in all sorts of crates (e.g. regex (?)).

**Comment by pwoolcoc** [comment]
I have a branch with this mostly working, I'm trying to get it cleaned up and committed "soon"

**Comment by tbu-** [comment]
Sounds great!

**Comment by marjakm** [comment]
Any updates?

**Comment by alexbool** [comment]
+1, would be great to have it

**Comment by Gyscos** [comment]
Also consider as possible filters for dependencies:
* Examples
* Integration tests
* Benchmarks

Also, should it be per-binary (when multiple are provided)?

**Comment by hauleth** [comment]
@Gyscos all of that can fall into `dev-dependencies` bucket. I see no reason for giving ability to set dependency only for benchmarks or something.

**Comment by azerupi** [comment]
@alexcrichton is this a desired feature? Would a PR proposing this get merged?

>Also, should it be per-binary (when multiple are provided)?

What is your opinion on this and how should it be represented in `Cargo.toml`?

**Comment by fulmicoton** [comment]
Would love to have this.

**Comment by alexcrichton** [comment]
@azerupi yeah I suspect Cargo will end up with something like this eventually, it's certainly a well motivated feature! I wouldn't got he route of dependencies-per-binary yet, though, as we don't similarly also have things like dependencies-per-example or dependencies-per-test

**Comment by alexreg** [comment]
Any progress on this yet?

**Comment by alexcrichton** [comment]
@alexreg unfortunately, no

This would likely require an RFC as it's an extension to Cargo.toml (e.g. a whole new section for dependencies)

**Comment by rminderhoud** [comment]
Hello, while this is being worked on is there a temporary "best practices" for handling this kind of project structure (root library crate -> member bin with additional dependencies)? Am I correct in saying that right now the only way to avoid polluting the library dependencies with the binary dependencies is to have each binary be a separate (non-workspaced) crate?

**Comment by alexcrichton** [comment]
@rminderhoud in the meantime the "best solution" (it's not really that great, but it works) is to split up into separate Cargo projects which can each have their own set of dependencies.

**Comment by steveklabnik** [comment]
Two recent issues where this has cropped up:

* https://github.com/steveklabnik/semver/pull/101
* https://github.com/google/pulldown-cmark/pull/55 and https://github.com/google/pulldown-cmark/pull/56

**Comment by crumblingstatue** [comment]
The most general solution would be to allow optional dependencies for every `bin` target. TOML makes this relatively easy to express:

```toml
[[bin]]
name = "foo"

[bin.dependencies]
clap = "1"
term = "0.5"

[[bin]]
name = "bar"

[bin.dependencies.cooldep]
version = "2.0"
default-features = false
features = ["foo"]
```

In the future, this general concept could also be expanded to examples/tests/etc., if deemed appropriate.

**Comment by alexreg** [comment]
I agree, the syntax & semantics proposed by @crumblingstatue look pretty sensible.

**Comment by alexcrichton** [comment]
cc https://github.com/rust-lang/cargo/pull/3870 a sample implementation of this.

**Comment by hobofan** [comment]
Another workaround I found to work well, is add a optinal feature `binaries`, and add a `required-features` field with that to the bin. As far as I know this hasn't been mentioned here before.

Example:
```toml
[[bin]]
name = "mybinary"
path = "src/main.rs"
required-features = ["binaries"]

[dependencies]
docopt = { version = "0.8.1", optional = true }

[features]
binaries = ["docopt"]
```

Trying to install the crate with `cargo install mycrate` fails with a pretty descriptive error, and as long as you document that it should be installed with `cargo install mycrate --features=binaries`, I don't really see a big downside to it.

A more consice way would of course be nice though ;)

**Comment by whitequark** [comment]
@hobofan this is sweet

**Comment by jonhoo** [comment]
@alexcrichton #3870 was closed with

> we may wish to continue to discuss this externally outside of a PR before resubmitting.

Has any more discussion taken place?

**Comment by alexcrichton** [comment]
Not a significant amount as far as I know at least

**Comment by jonhoo** [comment]
To build on this, it would be handy to have a way of declaring dependencies for only *some* binaries. For example, I have one binary that does EC2 orchestration, and depends on much of `rusoto`. It'd be sad to have to compile all of that if you're not running that binary.

@hobofan's workaround with `required-features` is nice, but it's inconvenient to use features for this as it also forces the library crate to be recompiled, which can take a non-trivial amount of time for larger crates. It of course also means that you need to add an additional flag for every binary you want to compile.

**Comment by alexreg** [comment]
This issue has been open some time now. Does anyone plan on tackling it?

**Comment by spease** [comment]
@crumblingstatue For consistency, wouldn't "[cooldep.bin.dependencies]" make more sense as it keeps the most specific item on the left, getting more general as you go to the right?

**Comment by crumblingstatue** [comment]
@stale bump.

There is real demand for this feature, so let's keep this open as a reminder that somebody should implement it someday.

**Comment by alexreg** [comment]
@crumblingstatue Looks like it will need RFC. Fancy writing one? I'd be glad to assist and review.

**Comment by crumblingstatue** [comment]
@alexreg Well, I do *want* this feature, but I'm not sure if I'm up to writing an RFC with a concrete implementation plan.

I think it's useful to track features that are in demand, but I understand if this issue tracker isn't for feature requests.

**Comment by hwchen** [comment]
@alexreg I'm interested in writing an RFC if you're still up for helping.

**Comment by alexreg** [comment]
@hwchen Great. And yes, I'd be glad to. Would you like to draft it then get my feedback, or discuss it on Discord/Zulip first? (That's where a lot of we Rust maintainers meet.)

**Comment by hwchen** [comment]
@alexreg awesome! let me read over this thread more carefully this weekend, then I'll ping you.

**Comment by OxyDeadbeef** [comment]
Is there an RFC for this now?

**Comment by alexreg** [comment]
@OxyDeadbeef Afraid not. @hwchen How are things with you? I forgot about this thread until now. Are you interested in drafting something up still, with my assistance?

**Comment by paulevans** [comment]
Is there anywhere I can be a fly on the wall for that so I can understand
how that process works?

**Comment by hwchen** [comment]
Sorry, I let this drop off. If anybody else would like to take this up
right now, please feel free.

**Comment by pwoolcoc** [comment]
Ok, I'm going to try and get a rough draft of an RFC to @alexreg in the next couple days

**Comment by alexreg** [comment]
@pwoolcoc That's super. I look forward to seeing it. If you have any questions in the meanwhile, tag me here or pop into the #design channel on Rust's Discord server.

**Comment by alexreg** [comment]
> Is there anywhere I can be a fly on the wall for that so I can understand how that process works?

@paulevans The process of RFC creation? The [README on the RFCs repo](https://github.com/rust-lang/rfcs/blob/master/README.md) provides a good overview, which @pwoolcoc will be following I'm sure. Usually the content of such RFC drafts is guided either my one's own ideas and musings or by pre-RFC discussion (over at internals.rust-lang.org). Sometimes they're a collaborative effort. Eventually an RFC PR is opened on the official Rust RFCs repo, when the community and in particular the Rust Lang (Design) team provides feedback on it. Usually the PR author(s) tag people who have been involved in the process, or expressed an interest. In this case we'll also post a link to the draft RFC here, so you can subscribe to that and follow the discussion and process.

**Comment by pwoolcoc** [comment]
@alexreg I am still working on this, just taking longer than I expected

**Comment by alexreg** [comment]
@pwoolcoc No problem. Here (or on Discord/Zulip) for advice if you need it, otherwise just ping me when you have a draft ready.

**Comment by Walther** [comment]
Excited Rust beginner here, also eager to try and help in any way i can. Writing a side project & bumped into this issue. I'm on both of the Discords with same username, feel free to ping! ❤️

**Comment by alexreg** [comment]
@Walther We'll a link to the RFC (draft) PR here when it's ready, which will hopefully be soon. :-)

**Comment by dmarcuse** [comment]
Has any progress been made with the RFC?

**Comment by alexreg** [comment]
@dmarcuse It has yes. If you'd be interested in taking it up, then great. I'm happy to provide advice or help review it.

**Comment by dmarcuse** [comment]
I'll start working on it this weekend and keep you updated!

**Comment by alexreg** [comment]
Super. You can contact me here, or on Discord/Zulip if you like.

**Comment by dmarcuse** [comment]
@alexreg I started drafting the RFC and sent you a friend request on Discord, could you accept it so I could ask some questions?

**Comment by alexreg** [comment]
@dmarcuse Okay thanks! Accepted now. (I think you can PM me regardless, but yeah, ask away...)

**Comment by djc** [comment]
@dmarcuse @alexreg have you made any progress since your last comments here?

**Comment by alexreg** [comment]
@djc Yep, @dmarcuse has begun a draft, and I've looked at a bit. Will review some more today.

**Comment by dmarcuse** [comment]
@djc I have been (very slowly) working on a draft available [here](https://github.com/dmarcuse/rfcs/blob/master/text/0000-cargo-target-dependencies.md) - feedback or contributions would be appreciated!

**Comment by djc** [comment]
@dmarcuse what you have so far looks good! For drawbacks and alternatives, it would be probably be useful to dig into what `require_features` currently does and how it would interact with your proposed feature. Otherwise this so far seems like a fairly straightforward feature and a mostly natural extension to what we already have, so there's probably not a lot of prior art and future possibilities to talk about.

My advice would be not to try to polish it all to perfection before you submit it, just write up whatever's in your mind and push it, then the community can help you with more feedback.

**Comment by dmarcuse** [comment]
Thanks for the tips @djc! I'm going to try to work on it more later today.

**Comment by alexreg** [comment]
Yes, we'll specifically get Alex Crichton's feedback on it once we have it in a decent state. :-)

**Comment by martingronlund** [comment]
What are the main reasons for not going with using a workspace or `dev-dependencies` here?

**Comment by jabedude** [comment]
Hello @dmarcuse! Any updates on this issue?

**Comment by dmarcuse** [comment]
I've been meaning to finish up and submit the RFC, but have been really busy with coursework lately - sorry for the delays. The current draft version is [available here](https://github.com/dmarcuse/rfcs/blob/master/text/0000-cargo-artifact-specific-dependencies.md) if you want to read through it, I'd appreciate feedback.

**Comment by linclelinkpart5** [comment]
Currently in this exact situation: I have a library, and a binary that uses said library. I'd prefer for them both to be in the same crate, but I also don't want the binary to depend on `clap`.

**Comment by est31** [comment]
There's an RFC now: https://github.com/rust-lang/rfcs/pull/2887

**Comment by pksunkara** [comment]
Created a new RFC for this, https://github.com/rust-lang/rfcs/pull/3020

**Comment by SoniEx2** [comment]
it'd be nice to have this as you can no longer rely on `extern crate` in edition 2018.

**Comment by Kixunil** [comment]
BTW, this affects bindgen - it optionally depends on `clap` because of binary but libraries using it in build scripts don't need `clap` and may forget to turn off default features.. This happened in case of `rocksdb` for instance.

**Comment by NobodyXu** [comment]
Perhaps we can enable all features specified by `required-features` in `[[bin]]` and `[[test]]`?

**Comment by epage** [comment]
https://github.com/rust-lang/cargo/issues/4663 is the issue for auto-enabling `required-features`, at least for examples with `cargo run`.

**Comment by charlesrocket** [comment]
Could we just tag dependencies with application scope when declaring them? So the compiler can do the rest:
```
[dependencies]
clap = { scope = ["bin"], version = "*"} 
rand = { scope = ["bin", "lib"], version = "*" }
rand_chacha = { scope = ["lib"], version = "*" }
```

**Comment by walksanatora** [comment]
Here to chime in with my 2 cents
Since we allready have 
`[target.*.dependicies]`
We could mabey have something like
`[target.bin.bin-target.dependicies]`
Or if that is not liked there is also the option for
`[[bin.bin-target.dependicies]]`
Either way this would make deploying
For Example
A client and a server from the same repo a little bit faster and easier to do

**Comment by zseri** [comment]
@walksanatora I think `[bin.dependencies]` (which will put a `dependency` table into the `[[bin]]` array-table) would be more appropriate.

**Comment by Mart-Bogdan** [comment]
@hauleth
> all of that can fall into `dev-dependencies` bucket. I see no reason for giving ability to set dependency only for benchmarks or something.

No, there is benefit -- reduce build time of tests, for example.
And reduce bloat and work on linker to strip uneded code. If bin would require test deps, we would need to buld all tests related stuff for binary.

**Comment by segevfiner** [comment]
I got hit by this when developing [segevfiner/keepawake-rs](https://github.com/segevfiner/keepawake-rs) (segevfiner/keepawake-rs#3), and I'm not sure how to restructure/split my crate in the mean time to avoid this issue. In terms of which part of the crate I should split into a sub-directory in a Cargo workspace and how to handle all the project metadata, README.md and so on, cause I'll have to write two, and/or how to correctly publish in such a configuration...

I tried the users forum but got no answer there as well https://users.rust-lang.org/t/whats-the-convention-for-handling-a-hybrid-library-and-binary-crates-dependencies/84174

**Comment by pksunkara** [comment]
New RFC has been proposed: https://github.com/rust-lang/rfcs/pull/3374

**Comment by detly** [comment]
For the specific use-case of binaries that exist to help with integration tests, I've made a [test-binary crate](https://crates.io/crates/test-binary) that might help. It's admittedly pretty niche, but it's why *I'm* watching this issue, so maybe it'll prove useful to others too.

**Comment by wiiznokes** [comment]
> New RFC has been proposed: https://github.com/rust-lang/rfcs/pull/3374

This is not the same feature of this issue

**Comment by mccolljr** [comment]
I hate to bump a stale thread, but I was curious about this today and google led me here. Is this a "dead" effort, or is there some blocker to implementing this that is not mentioned here?

**Comment by lure23** [comment]
I would use such feature to enable `defmt` feature (dependency) for all examples, while keeping it an opt-in for the `lib` itself.

**Comment by fmease** [comment]
With https://github.com/rust-lang/rfcs/pull/2887 officially closed, should this issue be closed as not planned? Just wondering.

**Comment by erhant** [comment]
any updates on this? I have the exact same use-case with the first comment!

**Comment by hukasu** [comment]
@pwoolcoc 10 year aniversary of your comment https://github.com/rust-lang/cargo/issues/1982#issuecomment-140190673 , working on Valve Time™, i see

**Comment by epage** [comment]
FYI this came up again on Internals: https://internals.rust-lang.org/t/breaking-down-crate-dependencies/23566

**Comment by matthieu-m** [comment]
To summarize the usecases I presented on Internals:

1. **Sans IO**: the ability to split the Sans IO core of a binary into a library:
    - Functionality: to ensure it's Sans IO, by ensuring it does not depend on any IO library (yes, I know, there's `std`...), but not `tokio` at least.
    - Compilation-time: to avoid dragging in the banana/gorilla/jungle in the component tests of the Sans IO core, which will mostly _emulate_ IO.

2. **Satellite Binaries**: the ability to provide a collection of "debugging" tools around the library, such as a streaming decoder binary for a protocol encoder/decoder library. For now, it's best to (ab)use examples for them, at the cost of bloating the doc-tests and unit-tests binaries with extra dependencies; at least it saves the "regular" downstream dependencies from that bloat.

3. **Dev Bloat**:
    - Unit-tests may need "mock" collections which benchmarks, examples, doc tests, or integration tests may not need.
    - Examples & Integration Tests may need actual IO (like tokio) which benchmarks & unit-tests may not need.
    - Benchmarks may need benchmark libraries (like criterion) which all others most likely do not need.

Spurious dependencies cause:

 1. Higher compile-times. For example, waiting on tokio-tls / rustls / aws-lc-rs / aws-lc-sys (Ouch) when compiling a unit-test which doesn't use TLS at all. This also has ripple effects on the downstream dependencies of spuriously pending dependencies.
 2. Larger binary sizes (and thus bloated target folders). While the symbols are tree-shaked, the debug info is not (apparently). This is especially relevant to the dev binaries (benchmarks, examples, doc tests, integration tests and unit-tests) which tend to accumulate every dependency and the kitchen sink, between spurious `[dev-dependencies]` and spurious binary dependencies.

**Comment by CrazySqueak** [comment]
Hi, are there any updates on this?

All of the available workarounds feel rather overcomplicated for something as simple as only wanting certain dependencies for certain binaries.

**Comment by zdivelbiss** [comment]
I'd like to voice my support for this feature, as well. I will work around this by separating the binary and library completely.
